feat: add gswrapper (GF-316) (#48)
GameFabric/docs: src/multiplayer-servers/multiplayer-services/game-server-wrapper.md

Before using the wrapper, make sure that:

- You have a container image that contains your game server binary
  (see [Building](/multiplayer-servers/getting-started/building-a-container-image) &
  [Pushing Container Images](/multiplayer-servers/getting-started/pushing-container-images)),
- You have either an Armada or Vessel configured to run the container image
  (see [Running your Game Server](/multiplayer-servers/getting-started/running-your-game-server)).

If your game server cannot run in GameFabric because it is missing vital information such as IP address or port numbers,
the wrapper can already help with its templating features.

## Integration

The wrapper is a binary executable file.
You have to run the wrapper instead of your game server, and the wrapper runs your game server,
including parameters or options.

Here are the steps to do this this:

- [Add the wrapper to your container image](#add-the-gsw-to-your-container-image)
    - Download the binary file,
    - Make it executable,
    - Build and push the new image.
- [Configure GameFabric](#configure-gamefabric)
    - Use the new image
    - Configure the wrapper

### Add the Wrapper to your Container Image

> Releases of the wrapper are publicly available here: https://github.com/gamefabric/gswrapper/releases.

Author: ```suggestion
Releases of the wrapper are publicly available at: https://github.com/gamefabric/gswrapper/releases.
```

# Game Server Wrapper

> The game server wrapper, also known as the wrapper, gswrapper or gsw, can be used to run your game server executable

Reviewer: ```
The game server wrapper, also known as `gswrapper` or `gsw`, can be used to launch your game server executable inside the container image that runs your GameFabric Armadas or Vessels.

By starting the executable as a wrapper it can enable various convenience features.
```

inside the container image that runs your GameFabric Armadas or Vessels.

The following features are available:

- [Parameter templating](#command-line-arguments)
- [Configuration file templating](#configuration-files),
- [Shutdown handling](#shutdown-handling),
- [Log tailing](#log-tailing), and
- [Crash reporting](#crash-reporting).

## Pre-requisites

Before using the wrapper, make sure that:

- You have a container image that contains your game server binary
  (see [Building](/multiplayer-servers/getting-started/building-a-container-image) &
  [Pushing Container Images](/multiplayer-servers/getting-started/pushing-container-images)),
- You have either an Armada or Vessel configured to run the container image
  (see [Running your Game Server](/multiplayer-servers/getting-started/running-your-game-server)).

If your game server cannot run in GameFabric because it is missing vital information such as IP address or port numbers,
the wrapper can already help with its templating features.

## Integration

The wrapper is a binary executable file.

> You have to run the wrapper instead of your game server, and the wrapper runs your game server,

Reviewer: ```
The wrapper is executed as the main command in the container and in turn executes your game server binary, passing it its command-line arguments.
```

including parameters or options.

Here are the steps to do this this:

- [Add the wrapper to your container image](#add-the-gsw-to-your-container-image)

>     - Download the binary file,

Reviewer: Remove commas?

    - Make it executable,
    - Build and push the new image.
- [Configure GameFabric](#configure-gamefabric)
    - Use the new image
    - Configure the wrapper

### Add the wrapper to your container image

> Releases of the wrapper are publicly available here: https://github.com/gamefabric/gswrapper/releases.

Reviewer: `available at`

Depending on how you create your container image, the integration for it can be as simple as:

```Dockerfile
ARG version=v2.2.0

ADD https://github.com/GameFabric/gswrapper/releases/download/${version}/gsw_linux_x86_64 \
    /app/gsw
RUN chmod +x /app/gsw
```

This downloads the x86 architecture build for Linux and makes it executable.

We provide builds for common used architectures and operating systems.
Contact us if you need a build that is not available.

::: info

> The wrapper is dependency free with one exception. It expects an Agones sidecar, which is always present in GameFabric.

Reviewer: ```diff
-It expects an Agones sidecar
+It communicates with the Agones sidecar container via localhost

> If you want to run it locally for development, you need run the Agones SDK server dummy locally (
> see [Agones documentation](https://agones.dev/site/docs/guides/client-sdks/local/)).

Author: ```suggestion
If you want to run it locally for development, you need run the Agones SDK server dummy locally
(see [Agones documentation](https://agones.dev/site/docs/guides/client-sdks/local/)).
```

:::

> We also follow [semantic versioning](https://semver.org/), which implies which version you can safely use or upgrade to without breaking changes.

Reviewer: ```suggestion
We use [semantic versioning](https://semver.org/), which indicates that minor and patch updates are safe to use without breaking changes.
```

> Finally, push your new image back into the GameFabric image registry, and make sure that the Armada or Vessel you're working with
> uses the updated image.

Reviewer: ```suggestion
Finally, push your new image to the GameFabric image registry, and ensure that the Armada or Vessel uses it.
```

> ### Configure GameFabric

Reviewer: Why does the example change `/bin/gameserver` into `gsw -- /bin/gameserver` when the UI looks like this?

![image](https://github.com/user-attachments/assets/bd652ce9-6037-4b6d-814c-dbd391e14385)

Author: Interesting. I always use just the arguments field. I don't see an advantage of specifying the command.

Whether you're using an Armada or a Vessel, configuring them to use the wrapper is the same.

> Go to Settings > Containers and update the new command for your game server container image.

Reviewer: ```suggestion
Go to Settings > Containers and update the command for your game server container image.
```

Before:

```shell
/app/gameserver <ARG> [--option]
```

After (with the wrapper):

```shell
/app/gsw -- /app/gameserver <ARG> [--option]
```

> This calls the wrapper instead of the game server (assuming `gameserver` this is the executable of your game server),

Reviewer: ```suggestion
This executes the wrapper instead of the game server (assuming `gameserver` this is the executable of your game server),
```

> but instructs the wrapper to call the game server with the arguments or options that your game server requires.

Reviewer: ```suggestion
but lets the wrapper pass the arguments on to your gameserver binary.
```

## Features

> The wrapper comes with a number of features to minimize your integration efforts with GameFabric or Agones,
> and to provide value around your stack, such as [log tailing](#log-tailing) or [crash reports](#crash-reporting).

Reviewer: ```suggestion
The wrapper provides a number of convenience features to facilitate the integration with GameFabric, such as [tailing of log files](#log-tailing) or [handling crashes and unclena exits](#crash-reporting).
```

### Templating

> The wrapper consumes basic information from the Agones Sidecar, like IP address or game port, as well as
> environment variables, e.g. set via GameFabric. This can also be implemented by your game server, and we encourage you to
> do so, but if you're not there yet, or if your game server relies on having a port passed to it
> you can access these variables through the wrapper.

Reviewer: ```suggestion
The wrapper collects basic runtime information from Agones, like IP address and ports, and the set of
environment variables for your gameserver.

It is able to pass the collected information as command-line arguments to your gameserver, or to render them into confugration files.

> [!TIP]
> The information collected is also available by directly querying the information via the Agones SDK.
```

> The wrapper uses the standard Go templating syntax and its basic functionality as
> described [here](https://pkg.go.dev/text/template#section-documentation).

Reviewer 2: ```suggestion
The wrapper uses the standard [Go templating syntax](https://pkg.go.dev/text/template#section-documentation) to configure the command line to run your game server.
```

IMO this is better here

> #### Command-line arguments

Reviewer 2: ```suggestion
#### Command-line Arguments
```

> The wrapper can be configured to generate a templated command to run the game server.

Reviewer 2: ```suggestion
```

And then we can remove this

| <code v-pre>{{.GameServerPort}}</code> | `uint16`            | The port given by Agones to the game server. Defaults to the port named `game`, or if there is no port with that name, to the first port found. A different name can be configured using `--ports.primary`. |
| <code v-pre>{{.Ports}}</code>          | `map[string]uint16` | The known game server ports.                                                                                                                                                                                |
| <code v-pre>{{.Env}}</code>            | `map[string]string` | The environment variables.                                                                                                                                                                                  |

**Example:**

```shell
/app/gsw -- /app/gameserver --port={{ .GameServerPort }} --query-port={{ .Ports.query }} --servername={{ .Env.POD_NAME }}
```

#### Configuration Files

> The wrapper can be configured to generate a template configuration file before invoking the game server.
> It needs a template file as input, which must be available in the wrapper's container, and an output file path.
> The available template variables are the same as for the [command-line arguments](#command-line-arguments).

Reviewer: ```suggestion
The wrapper can be configured to render a templated configuration file before executing the game server.

It needs a template file as input, which must be available in the wrapper's container, and an output file path.

The available template variables are the same as for the [command-line arguments](#command-line-arguments).
```

| Command-line argument    | Environment variable   | Description                                                |
|--------------------------|------------------------|------------------------------------------------------------|
| `--config.template-path` | `CONFIG_TEMPLATE_PATH` | Path at which to write the game server configuration file. |
| `--config.output-path`   | `CONFIG_OUTPUT_PATH`   | Path to the configuration file template.                   |

**Example:**

```yaml
# template.yaml
gameserver:
  ip: "{{ .GameServerIP }}"
  port: "{{ .GameServerPort }}"
  {{ - if .Env.POD_NAME }}
  servername: "{{ .Env.POD_NAME }}"
  {{ - end }}
```

```shell
gsw --config.template-path=template.yaml --config.output-path=config.yaml -- /app/gameserver --config=config.yaml
```

> ### Log tailing

Reviewer 2: ```suggestion
### Log Tailing
```

> The wrapper supports tailing log files and printing them to stdout using the wrapper's logger.

Reviewer: ```suggestion
The wrapper supports tailing log files and printing them to stdout using the wrapper's logger. This can be used to enable log collection for log files, which would otherwise be inaccessible.
```

> This allows you to see the logs in real time, for example if you're using a Vessel,
> and implies that they will be sent to our aggregated log solution.

Reviewer 2: ```suggestion
This allows you to see the logs in real time, for example if you're using a Vessel,
and implies that they get forwarded to our aggregated log solution.
```

No future tense please 😁

> This allows you to see the logs in real time, for example if you're using a Vessel,
> and implies that they will be sent to our aggregated log solution.

Reviewer: ```suggestion
In a containerized environment, only logs that are printed to stdout from the first process (PID 1) are collected and are available to be displayed and searched.
```

> This allows you to see the logs in real time, for example if you're using a Vessel,
> and implies that they will be sent to our aggregated log solution.

Author: ```suggestion
```

Removed after adding @towolf suggestion.

> This can be important because containers, including their log files, are volatile and would otherwise be lost if the container is stopped.

Reviewer: ```suggestion
Log files, on the other hand, would otherwise be lost as soon as the container of the gameserver is stopped.
```

| Command-line argument | Environment variable | Description                                                       |
|-----------------------|----------------------|-------------------------------------------------------------------|
| `--tail-log.paths`    | `TAIL_LOG_PATHS`     | Paths from which to tail log files. Can be passed multiple times. |

**Example:**

```shell
gsw --tail-log.paths=gameserver.log --tail-log.paths=error.log -- /app/gameserver
```

> ### Shutdown handling

Reviewer 2: ```suggestion
### Shutdown Handling
```

> The wrapper can manage the lifetime of the game server by initiating the shutdown for different phases and after a configured
> duration.

Reviewer: ```suggestion
The wrapper can terminate the game server after an elapsed amount of time, by shutting it down a configured duration, depending on it's state (`Scheduled`, `Ready`, `Allocated`).
```

> This can be useful to force the shutdown of stuck game servers or to allow fleet compaction.

Reviewer 2: ```suggestion
This is useful to force the shutdown of stuck game servers or to allow fleet compaction.
```

| Command-line argument  | Environment variable | Description                                                                                          |
|------------------------|----------------------|------------------------------------------------------------------------------------------------------|
| `--shutdown.scheduled` | `SHUTDOWN_SCHEDULED` | Shutdown when the game server has been `Scheduled` for the given duration (default: `0s`, disabled). |
| `--shutdown.ready`     | `SHUTDOWN_READY`     | Shutdown when the game server has been `Ready` for the given duration (default: `0s`, disabled).     |
| `--shutdown.allocated` | `SHUTDOWN_ALLOCATED` | Shutdown when the game server has been `Allocated` for the given duration (default: `0s`, disabled). |

`Scheduled`, `Ready` and `Allocated` refer to the Agones states explained [here](https://agones.dev/site/docs/reference/gameserver/#gameserver-state-diagram).

**Example:**

```shell
gsw --shutdown.ready=1h --shutdown.allocated=24h -- /app/gameserver
```

> ### Crash reporting

Reviewer 2: ```suggestion
### Crash Reporting
```

> Lastly, the crash handler can be configured to automatically run an executable in the event of a server crash.

Reviewer: ```suggestion
Lastly, the crash handler can be configured to run an executable in the event of a server crash. This executable could, for instance, run a debugger on the coredump file to generate a stack trace or to upload the full dump somewhere.
```

> The executable (e.g. a shell script to upload a crash dump) must be provided, executable and be part of the container image.

Reviewer: ```suggestion
The path to the executable must be specified, and the executable file itself must be present at the path in the image and carry the executable flag.
```

|-------------------------------------|-----------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
| `--crashhandler.exec`               | `CRASHHANDLER_EXEC`               | Path to the crash handler executable.                                                                                              |
| `--crashhandler.args`               | `CRASHHANDLER_ARGS`               | Crash handler arguments. Can be passed multiple times.                                                                             |
| `--crashhandler.max-execution-time` | `CRASHHANDLER_MAX_EXECUTION_TIME` | Timeout after which the crash handler should be aborted (default: `30m`). Please add the time unit, as the default is nanoseconds. |

**Example:**

```shell
gsw --crashhandler.exec=crash.sh --crashhandler.args="{{ .GameServerIP }}" --crashhandler.args="{{ .GameServerPort }}" --crashhandler.max-execution-time=5m
```

## Summary

> The wrapper extends the capabilities of your game server to facilitate interaction with Agones and within the Kubernetes infrastructure.
> You can choose to use some features while leaving others unused, and you can choose to integrate features into your game server yourself.
> Especially when gathering information about Agones, we encourage you to take the first steps as early as possible, as they will be needed later on anyway for an
> ideal integration.

Reviewer: ```suggestion
The wrapper provides a number of convenience features to facilitate the integration with GameFabric.
The offered features could all be implemented in the gameserver itself, however.
Especially using the Agones SDK to implement signaling of state and to obtain runtime information about your running Gameserver instances is recommended.
```
